### PR TITLE
enhance PythonPackage easyblock to deal with `posix_local` installation scheme used by Python in recent Debian/Ubuntu versions

### LINKED ISSUE #2976: Python packages are being installed to local/lib/python*/dist-packages due to tweaks to sysconfig.py in OS (Ubuntu 22.04)
**Problem description**

Hi. Recently, one of my machines started experiencing problems with installing EasyBuild via the bootstrap method.

```
== 2022-11-26 22:18:10,235 easyblock.py:3774 INFO Running method sanity_check_step part of step sanitycheck
== 2022-11-26 22:18:10,281 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/eb_tmp/local/lib/python3.10/dist-packages/easybuild/base/exceptions.py:124 in __init__): Failed to isolate installed egg dir for easybuild-framework (at easybuild/eb_tmp/local/lib/python3.10/dist-packages/easybuild/easyblocks/e/easybuildmeta.py:235 in sanity_check_step)
```

I use a custom build script I wrote based on the official instructions in the Lua, Lmod and EasyBuild documentations:
https://github.com/anselmicz/easybuild-installer/blob/main/easybuild.sh

My current builds fail during the sanity check phase, I'm enclosing the build log, environment at the time of installation, and trace output as a gist:
https://gist.github.com/anselmicz/72d1a0239cf67b685ed878aef2f61359

The error itself comes from https://github.com/boegel/easybuild-easyblocks/blob/develop/easybuild/easyblocks/e/easybuildmeta.py#L235

In a Slack conversation, @ocaisa mentioned that the error may be somehow connected to
https://github.com/easybuilders/easybuild-easyblocks/pull/2805

**System**

***OS***
Debian bookworm/sid

***Custom dependencies***
```
Lua/5.1.4.9
Lmod/8.5
```

***System packages***
```
python3/3.10.6
python3-pip/22.3
wget/1.21.3
gcc/12.2.0
make/4.3
rsync/3.2.6
tcl-dev/8.6.11
libreadline-dev/8.2
libibverbs-dev/43.0
xdot/1.2
```

***Python packages***
```
setuptools/65.6.3
python-graph-core/1.8.2
python-graph-dot/1.8.2
archspec/0.2.0
autopep8/2.0.0
GitPython/3.1.29
pep8/1.7.1
pycodestyle/2.10.0
Rich/12.6.0
```

Do you have any thoughts on this? I tried lowering both the temporary and to-be-installed versions of EasyBuild, but the issue persists both on `4.6.1` and  `4.6.0`, and I can't try `4.5.4` because then the `setuptools` and `setup.py` problems arise.

### PR DESCRIPTION
**Pull request by boegel**
fixes #2976

### CODE REVIEW

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _EasyBuild-4.8.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node4012.donphan.os - Linux Ubuntu 22.04.1 LTS (Jammy Jellyfish), x86_64, Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz (cascadelake), Python 3.10.6
See https://gist.github.com/boegel/0bbe4047ad6d144046c282740c93d802 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _EasyBuild-4.8.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
vpnz200.ugent.be - Darwin UNKNOWN UNKNOWN, AArch64, Apple M1 (m1), Python 3.9.6
See https://gist.github.com/boegel/7dfc46cfb795d3e1d5ed413711b8f0ea for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _EasyBuild-4.8.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3106.skitty.os - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.6.8
See https://gist.github.com/boegel/a668b898634f0bbf5da4e0a032d86721 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _Python-3.11.3-GCCcore-12.3.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node4012.donphan.os - Linux Ubuntu 22.04.1 LTS (Jammy Jellyfish), x86_64, Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz (cascadelake), Python 3.10.6
See https://gist.github.com/boegel/a2fa138d863eb40989cb6e4cd4962cc5 for a full test report.

File: easybuild/easyblocks/generic/pythonpackage.py [hunk 1]
```diff
@@ -224,6 +235,88 @@ def det_pip_version(python_cmd='python'):
...
+
+    install_dir_local = os.path.join(install_dir, 'local')
+    if os.path.exists(install_dir_local):
+        subdirs = os.listdir(install_dir)
+        log.info("Found 'local' subdirectory in installation prefix %s: %s", install_dir, subdirs)
+
+        local_subdirs = os.listdir(install_dir_local)
+        log.info("Subdirectories of %s: %s", install_dir_local, local_subdirs)
+
+        # symlink subdirectories of <prefix>/local to directly into <prefix>
```
Comment (jfgrimm) [inline easybuild/easyblocks/generic/pythonpackage.py:287]:
```suggestion
        # symlink subdirectories of <prefix>/local directly into <prefix>
```

File: easybuild/easyblocks/generic/pythonpackage.py [hunk 2]
```diff
@@ -539,6 +632,20 @@ def compose_install_command(self, prefix, extrapath=None, installopts=None):
 
         return ' '.join(cmd)
 
+    def py_post_install_shenanigans(self, install_dir):
+        """
+        Run post-installation shenanigans on specified installation directory, incl:
+        * dealing with 'local' subdirectory in install directory in case 'posix_local' installation scheme was used;
+        * symlinking site-packages to dist-packages is only the former is available;
```
Comment (jfgrimm) [inline easybuild/easyblocks/generic/pythonpackage.py:639]:
```suggestion
        * symlinking site-packages to dist-packages if only the former is available;
```

**Review by jfgrimm — CHANGES_REQUESTED** [review]

**Review by jfgrimm — APPROVED** [review]
lgtm

**Comment by jfgrimm** [comment]
Going in, thanks @boegel!

**Pull request merged**